tensorflow_io/ignite/kernels/ggfs/ggfs_client.cc

```diff
@@ -0,0 +1,377 @@
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  uint8_t res;
+  TF_RETURN_IF_ERROR(client_->ReadByte(&res));
+
+  if (!res)
+    return errors::NotFound("File ", path, " not found");
+
+  return Status::OK();
+}
+
+Status GGFSClient::Remove(string path) {
+  TF_RETURN_IF_ERROR(
+      SendCommonRequestHeader(kRemoveMethodId, 5 + path.length()));
+
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  return Status::OK();
+}
+
+Status GGFSClient::MkDir(string path, bool only_if_not_exists) {
+  TF_RETURN_IF_ERROR(
+      SendCommonRequestHeader(kMkDirMethodId, 6 + path.length()));
+
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteByte(only_if_not_exists));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  return Status::OK();
+}
+
+Status GGFSClient::MkDirs(string path, bool only_if_not_exists) {
+  TF_RETURN_IF_ERROR(
+      SendCommonRequestHeader(kMkDirsMethodId, 6 + path.length()));
+
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteByte(only_if_not_exists));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  return Status::OK();
+}
+
+Status GGFSClient::ListFiles(string path, std::vector<string> *out_files) {
+  TF_RETURN_IF_ERROR(
+      SendCommonRequestHeader(kListFilesMethodId, 5 + path.length()));
+
+  TF_RETURN_IF_ERROR(client_->WriteByte(kStringVal));
+  TF_RETURN_IF_ERROR(client_->WriteInt(path.length()));
+  TF_RETURN_IF_ERROR(client_->WriteData(
+      reinterpret_cast<const uint8_t *>(path.c_str()), path.length()));
+
+  TF_RETURN_IF_ERROR(ReceiveCommonResponseHeader());
+
+  int32_t length;
+  TF_RETURN_IF_ERROR(client_->ReadInt(&length));
+
+  while (length > 0) {
+    uint8_t type;
+    TF_RETURN_IF_ERROR(client_->ReadByte(&type));
+    if (type != kStringVal)
+      return errors::Unknown(
+          "Method GGFSClient::ListFiles expects strings in response");
+
+    int32_t str_length;
+    TF_RETURN_IF_ERROR(client_->ReadInt(&str_length));
+
+    uint8_t str[str_length];
+    TF_RETURN_IF_ERROR(client_->ReadData(str, str_length));
+
+    out_files->push_back(MakeRelative(
+        string(reinterpret_cast<char *>(str), str_length), path + "/"));
```

Comment:
are you sure that string c-tor will copy data from "str" to internal array? It don't remember...
because if it won't then data will be corrupted after method end

Reply:
Yep, I checked (see [here](http://www.cplusplus.com/reference/string/string/string/)).

```
(5) from buffer
    Copies the first n characters from the array of characters pointed by s.
```